<a href="https://colab.research.google.com/github/chaitanysaraswat/Advanced-Analytics-Using-Statistics/blob/main/lecture17.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# AUC-ROC
AREA UNDER CURVE AND RECIEVER OPERATING CHARACTERSTICS

In [1]:
from sklearn.metrics import roc_auc_score
y_true=[1,0,0,1,1,0]
y_pred=[0.93,0.83,0.31,0.73,0.25,0.37]
roc_auc_score(y_true,y_pred)

np.float64(0.5555555555555556)

#  NOMINAL LOGISTIC  REGRESSION:
 - IS USED WHEN YOUR RESPONSE IS MULTI-CATEGORICAL

In [2]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
import statsmodels.stats.multicomp
import sklearn
from sklearn.model_selection import train_test_split
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', None)
from numpy import random
from collections import Counter
from numpy.linalg import inv
from numpy.linalg import eig
import matplotlib
from matplotlib import pyplot as plt
import seaborn as sns
import pylab
from pylab import legend
from pylab import plot, show, title, xlabel, ylabel
import scipy
from scipy import stats
from scipy.stats import binom,poisson,norm,t,expon,f
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [4]:
df = pd.read_excel('/content/drive/MyDrive/CDAC_stats/CDAC_DataBook.xlsx', sheet_name='nominal')
df = df.drop('write', axis=1)

x = df.drop('prog', axis=1)
y = df.prog

ses_dummy = pd.get_dummies(x.ses, drop_first=True, prefix='SES').astype(int)
x.drop('ses', axis=1, inplace=True)
x = pd.concat([x,ses_dummy], axis=1)

x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2, random_state=20)

x_train = sm.add_constant(x_train, prepend=False)
mod1 = sm.MNLogit(y_train, x_train).fit() # multi nominal logical
print(mod1.summary())

Optimization terminated successfully.
         Current function value: 0.653667
         Iterations 8
                          MNLogit Regression Results                          
Dep. Variable:                   prog   No. Observations:                  160
Model:                        MNLogit   Df Residuals:                      152
Method:                           MLE   Df Model:                            6
Date:                Tue, 09 Jun 2026   Pseudo R-squ.:                  0.3687
Time:                        06:34:20   Log-Likelihood:                -104.59
converged:                       True   LL-Null:                       -165.68
Covariance Type:            nonrobust   LLR p-value:                 5.678e-24
    prog=2       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
math           0.0112      0.036      0.312      0.755      -0.059       0.082
SES_2          3.2174      0.

In [5]:
import statsmodels.formula.api as smf
from statsmodels.discrete.discrete_model import Poisson as psn
df = pd.read_excel('/content/drive/MyDrive/CDAC_stats/CDAC_DataBook.xlsx', sheet_name='poisson')
prog_dummy = pd.get_dummies(df.prog, drop_first=True,prefix='prog').astype(int)
df = df.drop('prog', axis=1)
df = pd.concat([df,prog_dummy], axis=1)
x_train, x_test, y_train, y_test = train_test_split(df.drop('num_awards', axis=1), df['num_awards'], test_size=0.2,random_state=20)
x_train1 = sm.add_constant(x_train, prepend=False)

df_train = pd.concat([x_train1,y_train], axis=1)

mod2 = psn.from_formula("num_awards ~ math+prog_2+prog_3", data=df_train).fit()

Optimization terminated successfully.
         Current function value: 0.871654
         Iterations 6


- null hypothesis 1: maths score doesnt influence the choice b/w program 1 and program2

- p=0.755

- does not reject

- influence :does not influence

- as math score increases probality of choosing prog2 will increase from prog1
---

- null hypothesis for ses2: whether we belong to ses1 or  ses 2, it does not influence choice b/w program 1 and program 2

- p=0

- reject hypothesis

- wrt  ses1 if ses 2selected then log of odd's ratio
---

- null hypothesis ses3: whether we belong to ses1 or  ses 3, it does not influence choice b/w program 1 and program 2

ORDINAL LOGISTIC REGRESSION:
- is used when there are response in multiple categories but are in logical order
- response ex- rating and ranking

COUNTS MODEL :
-  ARE USED WHEN THE RESPONSE IS STRICTLY A WHOLE NUMBER
- less variation (under dispersed) poison
- high variation (over dispersed) negative binomial

In [3]:
import statsmodels.formula.api as smf
from statsmodels.discrete.discrete_model import Poisson as psn
df = pd.read_excel('/content/drive/MyDrive/CDAC_stats/CDAC_DataBook.xlsx', sheet_name='poisson')
prog_dummy = pd.get_dummies(df.prog, drop_first=True,prefix='prog').astype(int)
df = df.drop('prog', axis=1)
df = pd.concat([df,prog_dummy], axis=1)
x_train, x_test, y_train, y_test = train_test_split(df.drop('num_awards', axis=1), df['num_awards'], test_size=0.2,random_state=20)
x_train1 = sm.add_constant(x_train, prepend=False)

df_train = pd.concat([x_train1,y_train], axis=1)

mod2 = psn.from_formula("num_awards ~ math+prog_2+prog_3", data=df_train).fit()

Optimization terminated successfully.
         Current function value: 0.871654
         Iterations 6
